[WIP] Minimal ePBS without Max EB and 7002 (#2)
potuz/consensus-specs: specs/_features/epbs/validator.md

    """
    Return ``GetInclusionListResponse`` object.
    """
    ...
```

Given the `parent_block_hash`, `get_execution_inclusion_list` returns `GetInclusionListResponse` with the most recent version of the inclusion list based on the parent block hash.

The body of this function is implementation dependent. The Engine API may be used to implement it with an external execution engine. This function returns a `GetInclusionListResponse` object that has to have the following properties
- The `transactions` list consists of a list of transactions that can be executed based on the current execution state parametrized by the `parent_block_hash`. This includes validations that each of the "from" addresses has enough funds to cover the maximum gas specified in the transactions at the specified `base_fee_per_gas` and any possible increase for the next block because of EIP-1559. It also includes checks for nonce correctness. 
- The total gas limit of all transactions in the list is less or equal than `MAX_GAS_PER_INCLUSION_LIST`. 
- The `summary` consists of the "from" addresses from the `transactions` list. In particular, these lists have the same length. 

## Validator assignment

> A validator can get PTC assignments for a given slot using the following helper via `get_ptc_assignment(state, epoch, validator_index)` where `epoch <= next_epoch`.

Reviewer: ```suggestion
A validator can get Payload Committee assignments for a given slot using the following helper via `get_payload_committee_assignment(state, epoch, validator_index)` where `epoch <= next_epoch`.
```

A validator can use the following function to see if they are supposed to submit payload attestation message during a slot across an epoch.

> PTC committee selection is only stable within the context of the current and next epoch.

Reviewer: ```suggestion
Payload Committee selection is only stable within the context of the current and next epoch.
```

```python

> def get_ptc_assignment(state: BeaconState,

Reviewer: ```suggestion
def get_payload_committee_assignment(state: BeaconState,
```

                             epoch: Epoch,
                             validator_index: ValidatorIndex
                             ) -> Optional[Slot]:
    """

>     Return the ptc committee assignment in the ``slot`` for ``validator_index``.

Reviewer: ```suggestion
    Return the payload committee assignment in the ``slot`` for ``validator_index``.
```

    ``assignment`` returned is a tuple of the following form:

>         * ``assignment[0]`` is the list of validators in the ptc

Reviewer: ```suggestion
        * ``assignment[0]`` is the list of validators in the payload committee
```

>         * ``assignment[1]`` is the slot at which the ptc is assigned

Reviewer: ```suggestion
        * ``assignment[1]`` is the slot at which the payload committee is assigned
```

    Return None if no assignment.
    """
    next_epoch = Epoch(get_current_epoch(state) + 1)
    assert epoch <= next_epoch

    start_slot = compute_start_slot_at_epoch(epoch)
    for slot in range(start_slot, start_slot + SLOTS_PER_EPOCH):

>             if validator_index in get_ptc(state, Slot(slot)):

Reviewer: ```suggestion
            if validator_index in get_payload_committee(state, Slot(slot)):
```

                return Slot(slot)
    return None
```

### Lookahead

[New in ePBS]

> `get_ptc_assignment` should be called at the start of each epoch to get the assignment for the next epoch (`current_epoch + 1`).

Reviewer: ```suggestion
`get_payload_committee_assignment` should be called at the start of each epoch to get the assignment for the next epoch (`current_epoch + 1`).
```

> A validator should plan for future assignments by noting their assigned ptc committee

Reviewer: ```suggestion
A validator should plan for future assignments by noting their assigned payload committee
```

> slot and planning to participate in the ptc committee subnet.

Reviewer: ```suggestion
slot and planning to participate in the payload committee subnet.
```

## Beacon chain responsibilities

All validator responsibilities remain unchanged other than the following:

- Proposers are required to broadcast `InclusionList` objects in addition to the `SignedBeaconBlock` objects. 
- Proposers are no longer required to broadcast `BlobSidecar` objects, as this becomes a builder's duty. 

> - Some validators are selected per slot to become PTC members, these validators must broadcat `PayloadAttestationMessage` objects

Reviewer: ```suggestion
- Some validators are selected per slot to become Payload Committee members, these validators must broadcat `PayloadAttestationMessage` objects
```

### Block proposal

Validators are still expected to propose `SignedBeaconBlock` at the beginning of any slot during which `is_proposer(state, validator_index)` returns `true`. The mechanism to prepare this beacon block and related sidecars differs from previous forks as follows

#### Inclusion lists generation
ePBS introduces forward inclusion lists. The proposer must construct and broadcast an `InclusionList` object alongside `SignedBeaconBlock`.

To obtain an inclusion list, a block proposer building a block on top of a `state` must take the following actions:

> 1. If `is_parent_block_full(state)` returns `False`, the proposer should not broadcast an inclusion list. Any inclusion list for this state will be ignored by validators. 

Reviewer: ```suggestion
1. If `is_parent_block_with_payload(state)` returns `False`, the proposer should not broadcast an inclusion list. Any inclusion list for this state will be ignored by validators. 
```

* The `signed_execution_payload_header` must satisfy the verification conditions found in `process_execution_payload_header`, that is 
    - The header signature must be valid
    - The builder balance can cover the header value
    - The header slot is for the proposal block slot
    - The header parent block hash equals the state's `latest_block_hash`. 
* Select one bid and set `body.signed_execution_payload_header = signed_execution_payload_header`

#### Constructing the new `payload_attestations` field in  `BeaconBlockBody`

Up to `MAX_PAYLOAD_ATTESTATIONS`, aggregate payload attestations can be included in the block. The validator will have to 
* Listen to the `payload_attestation_message` gossip global topic 
* The payload attestations added must satisfy the verification conditions found in payload attestation gossip validation and payload attestation processing. This means
    - The `data.beacon_block_root` corresponds to `block.parent_root`.
    - The slot of the parent block is exactly one slot before the proposing slot. 
    - The signature of the payload attestation data message verifies correctly. 

> * The proposer needs to aggregate all payload attestations with the same data into a given `PayloadAttestation` object. For this it needs to fill the `aggregation_bits` field by using the relative position of the validator indices with respect to the PTC that is obtained from `get_ptc(state, block_slot - 1)`. 

Reviewer: ```suggestion
* The proposer needs to aggregate all payload attestations with the same data into a given `PayloadAttestation` object. For this it needs to fill the `aggregation_bits` field by using the relative position of the validator indices with respect to the Payload Committee that is obtained from `get_payload_committee(state, block_slot - 1)`. 
```

* Proposer should only include payload attestations that are consistent with the current block they are proposing. That is, if the previous block had a payload, they should only include attestations with `payload_status = PAYLOAD_PRESENT`. Proposers are penalized for attestations that are not-consistent with their view. 

#### Blobs sidecars
The blbos sidecars are no longer broadcast by the validator, and thus their construction is not necessary. This deprecates the corresponding sections from the honest validator guide in the Deneb fork, moving them, albeit with some modifications, to the [honest Builder guide](./builder.md)

### Payload timeliness attestation

> Some validators are selected to submit payload timeliness attestation. Validators should call `get_ptc_assignment` at the beginning of an epoch to be prepared to submit their PTC attestations during the next epoch. 

Reviewer: ```suggestion
Some validators are selected to submit payload timeliness attestation. Validators should call `get_payload_committee_assignment` at the beginning of an epoch to be prepared to submit their Payload Committee attestations during the next epoch. 
```

    - If a valid `SignedExecutionPayloadEnvelope` has been seen referencing the block `data.beacon_block_root` and the envelope has `payload_withheld = True`,  set to `PAYLOAD_WITHHELD`.
    - If no valid `SignedExecutionPayloadEnvelope` has been seen referencing the block `data.beacon_block_root` set to `PAYLOAD_ABSENT`.
* Set `payload_attestation_message.validator_index = validator_index` where `validator_index` is the validator chosen to submit. The private key mapping to `state.validators[validator_index].pubkey` is used to sign the payload timeliness attestation.
* Sign the `payload_attestation_message.data` using the helper `get_payload_attestation_message_signature`.

Notice that the attester only signs the `PayloadAttestationData` and not the `validator_index` field in the message. Proposers need to aggregate these attestations as described above. 

```python
def get_payload_attestation_message_signature(state: BeaconState, attestation: PayloadAttestationMessage, privkey: int) -> BLSSignature:

>     domain = get_domain(state, DOMAIN_PTC_ATTESTER, compute_epoch_at_slot(attestation.data.slot))

Reviewer: ```suggestion
    domain = get_domain(state, DOMAIN_PAYLOAD_COMMITTEE_ATTESTER, compute_epoch_at_slot(attestation.data.slot))
```
